feat: split dashboards for 7mode - volume

Hardikl opened the pull request:
For 7mode system, vol create is based on aggr name, means aggr+vol name is unique. 
```
tobago-2> vol help create 
vol create <vol-name>
        { [-l <language-code>] [-s {none | file | volume}]
          <hosting-aggr-name> <size>[k|m|g|t] }
    |
        { [-l <language-code>] <hosting-aggr-name>
          [size[k|m|g|t]] -S <remotehost:remotevolume> }
    |
        { [-f] [-l <language-code>] [-m] [-n] 
          [-L [compliance | enterprise]]
          [-r <raid-group-size>] [-t {raid4 | raid_dp}]
          <disk-list> }
 -- create a new volume, either a flexible volume from an existing
   aggregate, or a traditional volume from a disk list.  A disk
   list is either
           <ndisks>[@<disk-size>] 
   or
           -d <disk-name1> <disk-name2> ... <disk-nameN>
           [-d <disk-name1> <disk-name2> ... <disk-nameN>].
```

link: https://library.netapp.com/ecm/ecm_download_file/ECMP1196890

Few examples from above link:

![image](https://user-images.githubusercontent.com/83282894/133063380-70ffb6b1-5ba9-4777-8133-20c6e50abb3e.png)



gaps:
- 7 mode perf counters don't have node/svm detail
   bin/zapi --poller cluster-04 show data --object volume | dasel -r xml -w json
...
...
          {
                "name": "node_name",
                "value": ""
              },
              {
                "name": "node_uuid",
                "value": ""
              },
              {
                "name": "process_name",
                "value": ""
              },
              {
                "name": "vserver_name",
                "value": ""
              },
              {
                "name": "vserver_uuid",
                "value": ""
              },
              {
...
...

As we are not able to get svm name detail in perf counters for 7mode, we use join query with volume labels such that it can be filterable in 7mode.
```
ex: volume_avg_latency{datacenter="$Datacenter",cluster="$Cluster",volume=~"$Volume"} * on  (volume, aggr) group_right() volume_labels{datacenter="$Datacenter",cluster="$Cluster",svm=~"$SVM",volume=~"$Volume"}
```

cgrinds reviewed (commented):
looks good @Hardikl

Hardikl commented:
for lun, 
lun-list-info zapi and lun counter don't have svm/vfiler detail, so svm level filter has been removed for lun dashboard in 7mode.

Hardikl commented:
Unit testing:
Import dashboards
Case 1:
fresh import:
bin/grafana import --addr localhost:3000 --directory grafana/dashboards
using API token from config
connected to Grafana server (version: 8.1.2)
created Grafana folder [Harvest 2.0] - OK
created Grafana folder [7 mode] - OK
preparing to import dashboards...
OK - imported [harvest_dashboard_aggregate7.json]
OK - imported [harvest_dashboard_cluster7.json]
OK - imported [harvest_dashboard_disk7.json]
OK - imported [harvest_dashboard_lun7.json]
OK - imported [harvest_dashboard_metadata7.json]
OK - imported [harvest_dashboard_network7.json]
OK - imported [harvest_dashboard_node7.json]
OK - imported [harvest_dashboard_nvme_fc7.json]
OK - imported [harvest_dashboard_shelf7.json]
OK - imported [harvest_dashboard_volume7.json]
Imported 10 dashboards from grafana/dashboards/7mode 
OK - imported [harvest_dashboard_aggregate.json]
OK - imported [harvest_dashboard_cluster.json]
OK - imported [harvest_dashboard_disk.json]
OK - imported [harvest_dashboard_lun.json]
OK - imported [harvest_dashboard_metadata.json]
OK - imported [harvest_dashboard_network.json]
OK - imported [harvest_dashboard_network_detail.json]
OK - imported [harvest_dashboard_node.json]
OK - imported [harvest_dashboard_node_details.json]
OK - imported [harvest_dashboard_nvme_fc.json]
OK - imported [harvest_dashboard_shelf.json]
OK - imported [harvest_dashboard_snapmirror.json]
OK - imported [harvest_dashboard_svm.json]
OK - imported [harvest_dashboard_svm_details.json]
OK - imported [harvest_dashboard_volume.json]
OK - imported [harvest_dashboard_volume_details.json]
Imported 16 dashboards from grafana/dashboards 

Case 2:
one folder exist other not(7 mode not exist):
hardikl@hardikl-mac-0 harvest % bin/grafana import --addr localhost:3000 --directory grafana/dashboards          
using API token from config
connected to Grafana server (version: 8.1.2)
folder [Harvest 2.0] exists in Grafana - OK
created Grafana folder [7 mode] - OK
preparing to import dashboards...
OK - imported [harvest_dashboard_aggregate7.json]
OK - imported [harvest_dashboard_cluster7.json]
OK - imported [harvest_dashboard_disk7.json]
OK - imported [harvest_dashboard_lun7.json]
OK - imported [harvest_dashboard_metadata7.json]
OK - imported [harvest_dashboard_network7.json]
OK - imported [harvest_dashboard_node7.json]
OK - imported [harvest_dashboard_nvme_fc7.json]
OK - imported [harvest_dashboard_shelf7.json]
OK - imported [harvest_dashboard_volume7.json]
Imported 10 dashboards from grafana/dashboards/7mode 
OK - imported [harvest_dashboard_aggregate.json]
OK - imported [harvest_dashboard_cluster.json]
OK - imported [harvest_dashboard_disk.json]
OK - imported [harvest_dashboard_lun.json]
OK - imported [harvest_dashboard_metadata.json]
OK - imported [harvest_dashboard_network.json]
OK - imported [harvest_dashboard_network_detail.json]
OK - imported [harvest_dashboard_node.json]
OK - imported [harvest_dashboard_node_details.json]
OK - imported [harvest_dashboard_nvme_fc.json]
OK - imported [harvest_dashboard_shelf.json]
OK - imported [harvest_dashboard_snapmirror.json]
OK - imported [harvest_dashboard_svm.json]
OK - imported [harvest_dashboard_svm_details.json]
OK - imported [harvest_dashboard_volume.json]
OK - imported [harvest_dashboard_volume_details.json]
Imported 16 dashboards from grafana/dashboards 


Case 3:
both folder exist(but removed 2-3 dashboards from grafana UI):
hardikl@hardikl-mac-0 harvest % bin/grafana import --addr localhost:3000 --directory grafana/dashboards
using API token from config
connected to Grafana server (version: 8.1.2)
folder [Harvest 2.0] exists in Grafana - OK
folder [7 mode] exists in Grafana - OK
preparing to import dashboards...
OK - imported [harvest_dashboard_aggregate7.json]
OK - imported [harvest_dashboard_cluster7.json]
OK - imported [harvest_dashboard_disk7.json]
OK - imported [harvest_dashboard_lun7.json]
OK - imported [harvest_dashboard_metadata7.json]
OK - imported [harvest_dashboard_network7.json]
OK - imported [harvest_dashboard_node7.json]
OK - imported [harvest_dashboard_nvme_fc7.json]
OK - imported [harvest_dashboard_shelf7.json]
OK - imported [harvest_dashboard_volume7.json]
Imported 10 dashboards from grafana/dashboards/7mode 
OK - imported [harvest_dashboard_aggregate.json]
OK - imported [harvest_dashboard_cluster.json]
OK - imported [harvest_dashboard_disk.json]
OK - imported [harvest_dashboard_lun.json]
OK - imported [harvest_dashboard_metadata.json]
OK - imported [harvest_dashboard_network.json]
OK - imported [harvest_dashboard_network_detail.json]
OK - imported [harvest_dashboard_node.json]
OK - imported [harvest_dashboard_node_details.json]
OK - imported [harvest_dashboard_nvme_fc.json]
OK - imported [harvest_dashboard_shelf.json]
OK - imported [harvest_dashboard_snapmirror.json]
OK - imported [harvest_dashboard_svm.json]
OK - imported [harvest_dashboard_svm_details.json]
OK - imported [harvest_dashboard_volume.json]
OK - imported [harvest_dashboard_volume_details.json]
Imported 16 dashboards from grafana/dashboards 


case 4:
custom names:
hardikl@hardikl-mac-0 harvest % bin/grafana import --addr localhost:3000 --directory grafana/dashboards -H cmode -f filer7
using API token from config
connected to Grafana server (version: 8.1.2)
created Grafana folder [cmode] - OK
created Grafana folder [filer7] - OK
preparing to import dashboards...
OK - imported [harvest_dashboard_aggregate7.json]
OK - imported [harvest_dashboard_cluster7.json]
OK - imported [harvest_dashboard_disk7.json]
OK - imported [harvest_dashboard_lun7.json]
OK - imported [harvest_dashboard_metadata7.json]
OK - imported [harvest_dashboard_network7.json]
OK - imported [harvest_dashboard_node7.json]
OK - imported [harvest_dashboard_nvme_fc7.json]
OK - imported [harvest_dashboard_shelf7.json]
OK - imported [harvest_dashboard_volume7.json]
Imported 10 dashboards from grafana/dashboards/7mode 
OK - imported [harvest_dashboard_aggregate.json]
OK - imported [harvest_dashboard_cluster.json]
OK - imported [harvest_dashboard_disk.json]
OK - imported [harvest_dashboard_lun.json]
OK - imported [harvest_dashboard_metadata.json]
OK - imported [harvest_dashboard_network.json]
OK - imported [harvest_dashboard_network_detail.json]
OK - imported [harvest_dashboard_node.json]
OK - imported [harvest_dashboard_node_details.json]
OK - imported [harvest_dashboard_nvme_fc.json]
OK - imported [harvest_dashboard_shelf.json]
OK - imported [harvest_dashboard_snapmirror.json]
OK - imported [harvest_dashboard_svm.json]
OK - imported [harvest_dashboard_svm_details.json]
OK - imported [harvest_dashboard_volume.json]
OK - imported [harvest_dashboard_volume_details.json]
Imported 16 dashboards from grafana/dashboards 


Export dashboards:
case 1: 
fresh export:
hardikl@hardikl-mac-0 harvest % bin/grafana export --addr localhost:3000 --directory grafana/A
using API token from config
connected to Grafana server (version: 8.1.2)
querying for content of folder id [222]
exporting dashboards to directory [grafana/A/Harvest_2.0]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_node.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_snapmirror.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_volume.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_volume_details.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_network_details.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_node_details.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_shelf.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_svm.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_svm_details.json]
OK - exported [grafana/A/Harvest_2.0/db_harvest_metadata.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_lun.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_network.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_network_with_nvme_fc.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_aggregate.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_cluster.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_disk.json]
exported 16 dashboards to [grafana/A/Harvest_2.0]
querying for content of folder id [223]
exporting dashboards to directory [grafana/A/7_mode]
OK - exported [grafana/A/7_mode/db_netapp_detail_shelf_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_aggregate_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_cluster_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_network_with_nvme_fc_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_node_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_volume_7_mode.json]
OK - exported [grafana/A/7_mode/db_harvest_metadata_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_disk_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_lun_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_network_7_mode.json]
exported 10 dashboards to [grafana/A/7_mode]


case 2:
one folder exist other not (7 mode deleted from grafana/A location):
hardikl@hardikl-mac-0 harvest % bin/grafana export --addr localhost:3000 --directory grafana/A
using API token from config
connected to Grafana server (version: 8.1.2)
querying for content of folder id [222]
exporting dashboards to directory [grafana/A/Harvest_2.0]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_network_details.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_node.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_node_details.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_aggregate.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_lun.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_snapmirror.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_volume_details.json]
OK - exported [grafana/A/Harvest_2.0/db_harvest_metadata.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_network.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_shelf.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_volume.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_cluster.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_network_with_nvme_fc.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_svm_details.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_disk.json]
OK - exported [grafana/A/Harvest_2.0/db_netapp_detail_svm.json]
exported 16 dashboards to [grafana/A/Harvest_2.0]
querying for content of folder id [223]
exporting dashboards to directory [grafana/A/7_mode]
OK - exported [grafana/A/7_mode/db_harvest_metadata_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_aggregate_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_node_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_volume_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_cluster_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_disk_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_lun_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_network_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_network_with_nvme_fc_7_mode.json]
OK - exported [grafana/A/7_mode/db_netapp_detail_shelf_7_mode.json]
exported 10 dashboards to [grafana/A/7_mode]


case 3: 
both folder exist in dir(but removed 2-3 dashboards from grafana/A location):
same as above

exported files at location grafana/A:
hardikl@hardikl-mac-0 harvest % find grafana/A -name "*.json" 
grafana/A/Harvest_2.0/db_netapp_detail_shelf.json
grafana/A/Harvest_2.0/db_netapp_detail_node.json
grafana/A/Harvest_2.0/db_netapp_detail_volume.json
grafana/A/Harvest_2.0/db_netapp_detail_disk.json
grafana/A/Harvest_2.0/db_netapp_detail_snapmirror.json
grafana/A/Harvest_2.0/db_netapp_detail_network.json
grafana/A/Harvest_2.0/db_netapp_detail_volume_details.json
grafana/A/Harvest_2.0/db_netapp_detail_cluster.json
grafana/A/Harvest_2.0/db_netapp_detail_lun.json
grafana/A/Harvest_2.0/db_netapp_detail_aggregate.json
grafana/A/Harvest_2.0/db_netapp_detail_svm_details.json
grafana/A/Harvest_2.0/db_netapp_detail_svm.json
grafana/A/Harvest_2.0/db_netapp_detail_network_with_nvme_fc.json
grafana/A/Harvest_2.0/db_netapp_detail_network_details.json
grafana/A/Harvest_2.0/db_netapp_detail_node_details.json
grafana/A/Harvest_2.0/db_harvest_metadata.json
grafana/A/7_mode/db_netapp_detail_shelf_7_mode.json
grafana/A/7_mode/db_harvest_metadata_7_mode.json
grafana/A/7_mode/db_netapp_detail_volume_7_mode.json
grafana/A/7_mode/db_netapp_detail_disk_7_mode.json
grafana/A/7_mode/db_netapp_detail_cluster_7_mode.json
grafana/A/7_mode/db_netapp_detail_node_7_mode.json
grafana/A/7_mode/db_netapp_detail_network_7_mode.json
grafana/A/7_mode/db_netapp_detail_aggregate_7_mode.json
grafana/A/7_mode/db_netapp_detail_network_with_nvme_fc_7_mode.json
grafana/A/7_mode/db_netapp_detail_lun_7_mode.json
hardikl@hardikl-mac-0 harvest %

rahulguptajss commented on cmd/tools/grafana/grafana.go:
```diff
@@ -196,69 +214,87 @@ func exportDashboards(opts *options) error {
-			//fmt.Printf("Skipping [%s]...\n", f.Name())
-			continue
+		// This is for 7 mode dashboards, for any new folder addition, this need to be followed
+		if f.IsDir() && (f.Name() == strings.ReplaceAll(grafana7modeFolderTitle, " ", "")) {
+			if err = importDashboards(opts, path.Join(dir, f.Name()), opts.folder[1].folderId); err != nil {
+				fmt.Println(err)
+				os.Exit(1)
+			}
+		} else {
+			importFiles(f, dir, folderId, &importedFiles)
 		}
```
may be just calling function with different argument based on if else would be easier than recursive call. This doesn't look good use of recursion to me.

Hardikl replied:
removed recursion from import

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -26,9 +26,10 @@ import (
-	grafanaDataSource  = "Prometheus"
+	clientTimeout           = 5
+	grafanaFolderTitle      = "Harvest 2.0"
+	grafana7modeFolderTitle = "7 mode"
```
would using the title `7-mode` simplify the code later that needs to remove the space?

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -54,14 +53,25 @@ type options struct {
 	useInsecureTLS bool
 }
 
+type Folder struct {
```
does this struct need to be exported?

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -41,9 +42,7 @@ type options struct {
-	folder         string // Grafana folder where to upload from where to download dashboards
-	folderId       int64
-	folderUid      string
+	folder         [2]Folder
```
would it be clearer if instead of assuming that `folder[0]` is cmode and `[1]` is 7mode to make that explicit with something like:
```go
cmodeFolder Folder
mode7Folder Folder
```

Hardikl replied:
done, Thanks.

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -73,30 +83,41 @@ func doImport(_ *cobra.Command, _ []string) {
+	checkAndCreateFolder()
+
+	fmt.Printf("preparing to import dashboards...\n")
+	// Recursive function call
```
old comment?

Hardikl replied:
removed comments

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -197,39 +215,43 @@ func exportDashboards(opts *options) error {
+	for _, file := range files {
+		if !strings.HasSuffix(file.Name(), ".json") {
+			//fmt.Printf("Skipping [%s]...\n", file.Name())
+			//fmt.Printf("preparing to import %d dashboards...\n", len(files))
```
old comment?

Hardikl replied:
removed

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -73,30 +83,41 @@ func doImport(_ *cobra.Command, _ []string) {
+	for i, exist := range folderExists {
+		if exist {
+			fmt.Printf("folder [%s] exists in Grafana - OK\n", opts.folder[i].folderName)
+		} else if err = createFolder(opts, i); err != nil {
```
`createFolder()` can be simplified to take one argument opts.folder[i]

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -523,45 +544,43 @@ func createFolder(opts *options) error {
-			//fmt.Println("FOUND FOLDER!")
-			//x.Print(0)
-			return true, nil
+		if x["title"].(string) == folder.folderName {
```
if title is missing from the map, the cast will cause this code to panic, might be better to check if title exists first using the ok idiom

Hardikl replied:
looks little odd, but still added checks for all 3 and it can pass only when dashboard have all 3 title, id and uid.

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -523,45 +544,43 @@ func createFolder(opts *options) error {
-			//x.Print(0)
-			return true, nil
+		if x["title"].(string) == folder.folderName {
+			folder.folderId = int64(x["id"].(float64))
```
both these lines may panic too

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -671,7 +690,8 @@ func init() {
 	Cmd.PersistentFlags().StringVarP(&opts.token, "token", "t", "", "API token issued by Grafana server for authentication")
 	Cmd.PersistentFlags().StringVarP(&opts.dir, "directory", "d", "grafana/dashboards/", "when importing, directory that contains dashboards.\nWhen exporting, directory to write dashboards to")
-	Cmd.PersistentFlags().StringVarP(&opts.folder, "folder", "f", grafanaFolderTitle, "Grafana folder name for the dashboards")
+	Cmd.PersistentFlags().StringVarP(&opts.folder[0].folderName, "harvestfolder", "H", grafanaFolderTitle, "Grafana folder name for the C mode dashboards")
```
don't we want the old api to still work, meaning `--folder` and `-f` should still point to the c-mode dashboard

Hardikl replied:
yes, it should be.

cgrinds commented on cmd/tools/grafana/grafana.go:
```diff
@@ -671,7 +690,8 @@ func init() {
 	Cmd.PersistentFlags().StringVarP(&opts.dir, "directory", "d", "grafana/dashboards/", "when importing, directory that contains dashboards.\nWhen exporting, directory to write dashboards to")
-	Cmd.PersistentFlags().StringVarP(&opts.folder, "folder", "f", grafanaFolderTitle, "Grafana folder name for the dashboards")
+	Cmd.PersistentFlags().StringVarP(&opts.folder[0].folderName, "harvestfolder", "H", grafanaFolderTitle, "Grafana folder name for the C mode dashboards")
+	Cmd.PersistentFlags().StringVarP(&opts.folder[1].folderName, "7modefolder", "f", grafana7modeFolderTitle, "Grafana folder name for the 7 mode dashboards")
```
maybe `--folder-7mode` and no short option?

Hardikl replied:
done

cgrinds reviewed (commented):
nice job @Hardikl

Hardikl commented:
Covered panels from details page of volume, node and network and validated against tobago-1 and tobago-2.

Removed panels:
Node fan status
node power status
avg throughput
read write throughput
fcp front drill (5) because fcp_lif not exist
nvme/fc front drill(5) because nvme_lif not exist
throughput by lif
top volume end-to-end qos drilldown(8) because qos_volume_xxx not exist
top volume qos resource latency drilldown(9) because qos_detail_volume_xxx not exist
per volume snapshot space used
per volume snapshot space used perc
top volume fabric pool -> it is empty title

Added panels:
Node uptime
Ethernet port utilization percent
FCP port utilization perc
Ethernet port throughput 
avg read latency
avg latency
avg write latency
avg iops
read write latency
read write iops
icsci latency
icsci iops
avg latency by lif
iops by lif
ethernet throuput
per volume space used
per volume space used perc
volumes in cluster
top volume read latency
top volume write latency
top volume read throughput
top volume write throughput
top volume read iops
top volume write iops


updated: 
FCP port throughput
fcp throughput
fcp read th  fcp send
fcp write th  fcp receive

No change: (both same)
volume wafl layer - backend wafl drill

cgrinds reviewed (approved):
looks great
